### PR TITLE
Refactor chat component

### PR DESCRIPTION
**Pull request by narender2031**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
- Refactor chat component. 

## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes, as well
as any relevant images for UI changes._

## Added tests?

- [ ] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.forem.com
- [ ] readme
- [ ] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

**Comment by reobin** [comment]
Hey @narender2031 !

Saw "refactor chat component", and it caught my attention. I just fixed a bug with the chat component causing it to not clear the input after being sent. It's over at #10487. 

I think it would be compatible with your work if you want to take it into consideration while refactoring, especially the stuff I did with the Compose and Chat components, which cleans up the chat textarea's input state management.

I also added tests for the textarea being cleared after the message being sent, which can help you make sure nothing went wrong with the merge.

File: app/javascript/chat/ActiveChatChannel/IntroductionMessages/OpenIntroMessage.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { h } from 'preact';
+import PropTypes from 'prop-types';
+
+const OpenInroMessage = ({ activeChannel }) => {
```
Comment (citizen428) [inline app/javascript/chat/ActiveChatChannel/IntroductionMessages/OpenIntroMessage.js:4]: This is a typo, it should be `Intro`, not `Inro`.

File: app/javascript/chat/ActiveChatChannel/IntroductionMessages/DireactIntoMessage.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { h } from 'preact';
+import PropTypes from 'prop-types';
+
+const DireactIntroMessages = ({ activeChannel }) => {
```
Comment (citizen428) [inline app/javascript/chat/ActiveChatChannel/IntroductionMessages/DireactIntoMessage.js:4]: This is a typo: `Direct`, not `Direact`.

**Comment by nickytonline** [comment]
@narender2031, I'm going to take a look at this in tomorrow morning EDT before our meeting.

File: app/javascript/chat/ActiveChatChannel/ChatMessages.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  messages,
+  showTimestamp,
+  activeChannel,
+  currentUserId,
+  triggerActiveContent,
+  triggerEditMessage,
+  triggerDeleteMessage,
+}) => {
+  if (!messages[activeChannelId]) {
+    return '';
```
Comment (nickytonline) [inline app/javascript/chat/ActiveChatChannel/ChatMessages.js:19]: refactor (best practice): There is no need to return an empty string. If a component ends up returning nothing via conditional rendering, return `null`.

File: app/javascript/chat/ActiveChatChannel/ChatMessages.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        onDeleteMessageTrigger={triggerDeleteMessage}
+        onEditMessageTrigger={triggerEditMessage}
+      />
+    ),
+  );
+};
+
+Message.propTypes = {
+  activeChannelId: PropTypes.number,
+  messages: PropTypes.arrayOf(PropTypes.object),
```
Comment (nickytonline) [inline app/javascript/chat/ActiveChatChannel/ChatMessages.js:64]: refactor (non-blocking): Consider using `PropTypes.shape` to describe the shape of a `message` prop. If the message proptype exist elsewhere, consider importing a shred prop type. For an example of this, see https://github.com/forem/forem/blob/master/app/javascript/common-prop-types/article-prop-types.js.

For more information on `PropTypes.shape`, see the [PropTypes](https://reactjs.org/docs/typechecking-with-proptypes.html) documentation.

File: app/javascript/chat/ActiveChatChannel/IntroductionMessages/OpenIntroMessage.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  return (
+    <div className="chatmessage" style={{ color: 'grey' }}>
+      <div className="chatmessage__body">
+        You have joined {activeChannel.channel_name}! All interactions{' '}
```
Comment (nickytonline) [inline app/javascript/chat/ActiveChatChannel/IntroductionMessages/OpenIntroMessage.js:8]: refactor (non-blocking): Consider using a [template string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) instead.

```suggestion
        {`You have joined ${activeChannel.channel_name}! All interactions `}
```

File: app/javascript/chat/actions/actions.js [hunk 6]
```diff
@@ -1,15 +1,16 @@
...
-  fetch(`/chat_channels/${channelId}?message_offset=${messageOffset}`, {
-    Accept: 'application/json',
-    'Content-Type': 'application/json',
-    credentials: 'same-origin',
-  })
-    .then((response) => response.json())
-    .then(successCb)
-    .catch(failureCb);
-}
+// export function getAllMessages(channelId, messageOffset, successCb, failureCb) {
```
Comment (nickytonline) [inline app/javascript/chat/actions/actions.js:4]: refactor (blocking): Remove dead commented out code.

File: app/javascript/chat/actions/actions.js [hunk 7]
```diff
@@ -56,20 +57,30 @@ export function editMessage(editedMessage, successCb, failureCb) {
     .catch(failureCb);
 }
 
-export function sendOpen(activeChannelId, successCb, failureCb) {
-  fetch(`/chat_channels/${activeChannelId}/open`, {
+// export function sendOpen(activeChannelId, successCb, failureCb) {
```
Comment (nickytonline) [inline app/javascript/chat/actions/actions.js:60]: refactor (blocking): remove dead commented out code.

File: app/javascript/chat/actions/actions.js [hunk 8]
```diff
@@ -1,15 +1,16 @@
...
+//     Accept: 'application/json',
+//     'Content-Type': 'application/json',
+//     credentials: 'same-origin',
+//   })
+//     .then((response) => response.json())
+//     .then(successCb)
+//     .catch(failureCb);
+// }
 
 export function sendMessage(messageObject, successCb, failureCb) {
```
Comment (nickytonline) [inline app/javascript/chat/actions/actions.js:15]: refactor (non-blocking): Consider making this an async function as it makes it clearer to a consumer of the function that a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) is being returned.

```suggestion
export async function sendMessage(messageObject, successCb, failureCb) {
```

File: app/javascript/chat/actions/actions.js [hunk 9]
```diff
@@ -56,20 +57,30 @@ export function editMessage(editedMessage, successCb, failureCb) {
...
+//     },
+//     body: JSON.stringify({}),
+//     credentials: 'same-origin',
+//   })
+//     .then((response) => response.json())
+//     .then(successCb)
+//     .catch(failureCb);
+// }
+
+export async function sendOpen(activeChannelId) {
```
Comment (nickytonline) [inline app/javascript/chat/actions/actions.js:76]: refactor (non-blocking): It's not clear to me what `sendOpen` means. Consider a more descriptive name explaining the intent of the function.

File: app/javascript/chat/actions/actions.js [hunk 10]
```diff
@@ -56,20 +57,30 @@ export function editMessage(editedMessage, successCb, failureCb) {
...
-      'X-CSRF-Token': window.csrfToken,
-      'Content-Type': 'application/json',
-    },
-    body: JSON.stringify({}),
     credentials: 'same-origin',
-  })
-    .then((response) => response.json())
-    .then(successCb)
-    .catch(failureCb);
+    body: JSON.stringify({}),
```
Comment (nickytonline) [inline app/javascript/chat/actions/actions.js:80]: The `request` utility function converts the body automatically to JSON. See https://github.com/forem/forem/blob/master/app/javascript/utilities/http/request.js#L35

```suggestion
    body: {},
```

File: app/javascript/chat/actions/actions.js [hunk 11]
```diff
@@ -129,25 +140,39 @@ export function getChannels(
...
     credentials: 'same-origin',
-  })
-    .then((response) => response.json())
-    .then((json) => {
-      successCb(json.unopened_ids);
-    });
+  });
+
+  return response.json()
+  // fetch('/chat_channels?state=unopened_ids', {
```
Comment (nickytonline) [inline app/javascript/chat/actions/actions.js:149]: refactor (blocking): remove dead commented out code.

File: app/javascript/chat/actions/actions.js [hunk 12]
```diff
@@ -129,25 +140,39 @@ export function getChannels(
...
+//     'Content-Type': 'application/json',
+//     credentials: 'same-origin',
+//   })
+//     .then((response) => response.json())
+//     .then(successCb)
+//     .catch(failureCb);
+// }
+
+export async function getContent(url) {
+  const response = await request(url, {
```
Comment (nickytonline) [inline app/javascript/chat/actions/actions.js:170]: `'GET'` is the default and `credentials: 'same-origin'` are included automatically when using the `request` utility function. See lines https://github.com/forem/forem/blob/master/app/javascript/utilities/http/request.js#L27 and  https://github.com/forem/forem/blob/master/app/javascript/utilities/http/request.js#L47

File: app/javascript/chat/actions/actions.js [hunk 13]
```diff
@@ -180,3 +205,21 @@ export function deleteMessage(messageId, successCb, failureCb) {
...
+
+
+export async function getAllMessages(channelId, messageOffset) {
+  const response = await request(`/chat_channels/${channelId}?message_offset=${messageOffset}`, {
+    method: 'GET',
+    credentials: 'same-origin',
+  })
+  
+  return response.json();
+  // fetch(`/chat_channels/${channelId}?message_offset=${messageOffset}`, {
```
Comment (nickytonline) [inline app/javascript/chat/actions/actions.js:217]: refactor (blocking): remove dead commented out code.

File: app/javascript/chat/actions/actions.js [hunk 14]
```diff
@@ -180,3 +205,21 @@ export function deleteMessage(messageId, successCb, failureCb) {
     .then(successCb)
     .catch(failureCb);
 }
+
+
+export async function getAllMessages(channelId, messageOffset) {
+  const response = await request(`/chat_channels/${channelId}?message_offset=${messageOffset}`, {
+    method: 'GET',
```
Comment (nickytonline) [inline app/javascript/chat/actions/actions.js:212]: `'GET'` is the default and `credentials: 'same-origin'` are included automatically when using the `request` utility function. See lines https://github.com/forem/forem/blob/master/app/javascript/utilities/http/request.js#L27 and  https://github.com/forem/forem/blob/master/app/javascript/utilities/http/request.js#L47

**Review by nickytonline — CHANGES_REQUESTED** [review]
@narender2031  the draft is starting to look good. I have some feedback and request changes. I'll continue reviewing in the morning.

**Review by nickytonline — COMMENTED** [review]
refactor (non-blocking): All components and variables defined with an arrow function are unnecessary since there is no use of `this`. Prefer using a traditional function. This is something I need to update in our frontend documentation and potentially incorporate a lint rule.

For more information about this see the official [arrow function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) documentation on MDN as well as this post [When (and why) you should use ES6 arrow functions — and when you shouldn’t](https://www.freecodecamp.org/news/when-and-why-you-should-use-es6-arrow-functions-and-when-you-shouldnt-3d851d7f0b26).

File: app/javascript/chat/connectReducer.js [hunk 15]
```diff
@@ -0,0 +1,307 @@
+export function connectReducer(state, action) {
```
Comment (nickytonline) [inline app/javascript/chat/connectReducer.js:1]: documentation (blocking): For any functions, constants etc. that are exported in an ES module, provide documentation using JSDoc comments. For an example, see https://github.com/forem/forem/blob/master/app/javascript/utilities/http/request.js#L1.

**Review by nickytonline — CHANGES_REQUESTED** [review]

File: app/javascript/chat/components/ConnectStateProvider.jsx [hunk 16]
```diff
@@ -0,0 +1,22 @@
+import { h, createContext } from 'preact';
```
Comment (nickytonline) [inline app/javascript/chat/components/ConnectStateProvider.jsx:1]: @narender2031, @citizen428, since we're going to go with useReducer and context, I'll merge my PR, #9799, first.

Comment (narender2031) [reply app/javascript/chat/components/ConnectStateProvider.jsx:1]: sure

**Comment by nickytonline** [comment]
@narender2031, @citizen428, once this PR is out of draft mode, I will give it another review.

File: app/javascript/chat/components/ConnectStateProvider.jsx [hunk 17]
```diff
@@ -12,10 +12,9 @@ const { Provider } = store;
  * @param {object} props.initialState The initial value to store as state
  * @param {Function} props.reducer Given the current state and an action, returns the new state
  */
-function ConnectStateProvider({ children, initialState, reducer }) {
-  const [state, dispatch] = useReducer(reducer, initialState);
 
+const ConnectStateProvider = ({ children, initialState, reducer }) => {
```
Comment (nickytonline) [inline app/javascript/chat/components/ConnectStateProvider.jsx:16]: This change isn't necessary. It's fine that it uses the function keyword instead of an arrow function. Arrow functions work well when you need scope, i.e. "this".

File: app/javascript/chat/connectReducer.js [hunk 18]
```diff
@@ -0,0 +1,320 @@
+import * as Type from './components/ChatTypes';
+export function connectReducer(state, action) {
+  const { type, payload = {} } = action;
```
Comment (nickytonline) [inline app/javascript/chat/connectReducer.js:3]: refactor (blocking): There is no need to give payload a default value of `{}`. This will actually mask potential errors when a payload isn't being sent that may be required by an action.

File: app/javascript/chat/connectReducer.js [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+
+  switch (type) {
```
Comment (nickytonline) [inline app/javascript/chat/connectReducer.js:5]: refactor (non-blocking): consider combining reducers instead of a super long switch case.

**Review by nickytonline — CHANGES_REQUESTED** [review]
Some comments and request changes.

**Review by nickytonline — COMMENTED** [review]
I mentioned the ones about converting `<div />`s to actual buttons via our `<Button />` component, but I noticed there are a lot of `<div />`s being used instead of potentially better matched semantic HTML elements.

File: spec/system/admin/admin_creates_new_page.rb [hunk 20]
```diff
@@ -0,0 +1,23 @@
+require "rails_helper"
+
+RSpec.describe "Admin creates new page", type: :system do
```
Comment (citizen428) [inline spec/system/admin/admin_creates_new_page.rb:3]: I'm a bit confused, is this spec really related to the changes in this PR?

Comment (narender2031) [reply spec/system/admin/admin_creates_new_page.rb:3]: these changes updated due to conflict. I think this file is deleted in the master branch. I will check and update this.

File: app/javascript/chat/ActiveChatChannel/ActiveChannelMemberList.js [hunk 21]
```diff
@@ -1,6 +1,31 @@
 import { h } from 'preact';
 import PropTypes from 'prop-types';
 
+/**
+ * This component is used to render the list of all Achtive chat channel Members
```
Comment (nickytonline) [inline app/javascript/chat/ActiveChatChannel/ActiveChannelMemberList.js:5]: spelling (nitpick)

```suggestion
 * This component is used to render the list of all Active chat channel Members
```

File: app/javascript/chat/ActiveChatChannel/ChatMessages.js [hunk 22]
```diff
@@ -3,7 +3,36 @@ import PropTypes from 'prop-types';
 import ActionMessage from '../actionMessage';
 import Message from '../message';
 import DirectChatInfoMessage from './IntroductionMessages/DireactIntroMessages';
-import OpenChatInfoMessagge from './IntroductionMessages/OpenItnroMessage';
+import IntroductionMessage from './IntroductionMessages/OpenItnroMessage';
```
Comment (nickytonline) [inline app/javascript/chat/ActiveChatChannel/ChatMessages.js:6]: spelling (nitpick): OpenItnroMessage should be OpenIntroMessage

```suggestion
import IntroductionMessage from './IntroductionMessages/OpenIntroMessage';
```

**Review by citizen428 — APPROVED** [review]
I _think_ all the feedback has been addressed now, thanks @narender2031! :rocket:

File: app/javascript/chat/connectReducer.js (hunk 19 above)
Comment (narender2031) [reply app/javascript/chat/connectReducer.js:14]: I will do these changes incoming PRs with more refactoring.

**Comment by narender2031** [comment]
> I _think_ all the feedback has been addressed now, thanks @narender2031! 🚀

@citizen428 & @nickytonline  I need a small help Can you please test these changes locally. Just for a safe case.

**Comment by citizen428** [comment]
@narender2031 Could you please fill in the "QA Instructions" section of the PR to point out what to look for specifically? This also helps other potential reviewers who weren't following along.

**Comment by narender2031** [comment]
> @narender2031 Could you please fill in the "QA Instructions" section of the PR to point out what to look for specifically? This also helps other potential reviewers who weren't following along.

@citizen428 & @nickytonline 

QA Instructions:-
- All Messages in the chat channel should load.
- You can delete, update, the message from the chat channel.
- you can send the messages.
- you can open the chat channel settings and perform the actions.
- you can switch the chat channels.
- you can open the mention POP up (*NOTE- Mention is not working in this PR. Another PR with the FIX is already done)
- you can search the chat channels.
- you can open the request center

**Comment by citizen428** [comment]
Quick question: since this is a refactoring PR, are you planning on also addressing some of the CodeClimate issues or will you do that separately?

**Comment by narender2031** [comment]
> Quick question: since this is a refactoring PR, are you planning on also addressing some of the CodeClimate issues or will you do that separately?

we will do separately because some of the files have a large number of the line so we will refactor those also one by one.

**Comment by citizen428** [comment]
Feedback: 

1. On first starting the app after checking out this branch I get the following when opening the site in the browser:
    > Failed to compile.
    >
    > ./app/javascript/chat/channels.jsx
    > Module not found: Error: [CaseSensitivePathsPlugin] `/Users/xxx/src/work/forem/app/javascript/chat/components/ChannelButton.jsx` does not match the corresponding path on disk `channelButton.jsx`.

    Renaming the file  fixed this:
    ```
    $ git mv app/javascript/chat/components/{c,C}hannelButton.jsx
    ```
2. The placeholder text says "Write message to undefined"
    Screen Shot 2020-11-10 at 13 13 24
3. It seems messages don't show up for me until I refresh. That's true for both new messages and edited ones as well as when deleting messages:

**Comment by narender2031** [comment]
> Feedback:
> 3. It seems messages don't show up for me until I refresh. That's true for both new messages and edited ones as well as when deleting messages:

@citizen428  You need to integrate the PUSHER locally for this. Can you please check with the other branch? is that happening with all. Because in my case locally in other branches messages don't show up on send or updates. It requires refresh.

**Comment by citizen428** [comment]
Yes @narender2031, correct re 3, I just realized this after changing branches and noticing the same behavior. Sorry for the false alarm on this one.

**Comment by narender2031** [comment]
> Yes @narender2031, correct re 3, I just realized this after changing branches and noticing the same behavior. Sorry for the false alarm on this one.

np @citizen428

**Comment by narender2031** [comment]
> Feedback:
> 
> 1. On first starting the app after checking out this branch I get the following when opening the site in the browser:

> 2. The placeholder text says "Write message to undefined"



@citizen428  issues are fixed.

**Comment by nickytonline** [comment]
@narender2031 and @sarthology, can we pause work on this PR until #11332 and #9741 are finished? There are change requests in both of those PRs. cc: @citizen428

**Comment by nickytonline** [comment]
As discussed, going to put this on hold and reassess the refactor. Closing this for now.
